hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java

```diff
@@ -255,11 +489,210 @@ public Duration timeBefore2ExpiryGracePeriod(X509Certificate certificate) {
+              " time root rotation {} is failed. Release the lock.",
+              rootCACertId);
+          releaseLock();
+        }, ackTimeout.toMillis(), TimeUnit.MILLISECONDS);
+
+      } catch (Throwable e) {
+        LOG.error("Unexpected error happen", e);
+        scm.shutDown("Unexpected error happen, " + e.getMessage());
+      }
+    }
+  }
+
+  /**
+   *  Task to wait the all acks of prepare request.
+   */
+  public class WaitSubCARotationPrepareAckTask implements Runnable {
+    private String rootCACertId;
+
+    public WaitSubCARotationPrepareAckTask(
+        X509CertificateHolder rootCertHolder) {
+      this.rootCACertId = rootCertHolder.getSerialNumber().toString();
+    }
+
+    @Override
+    public void run() {
+      if (!isRunning()) {
+        LOG.info("SCM is not leader anymore. Delete the in-progress " +
+            "root CA directory");
+        cleanupAndStop("SCM is not leader anymore");
+        return;
+      }
+      synchronized (RootCARotationManager.class) {
+        int numFromHADetails =
+            scm.getSCMHANodeDetails().getPeerNodeDetails().size() + 1;
+        int numFromRatisServer = scm.getScmHAManager().getRatisServer()
+            .getDivision().getRaftConf().getCurrentPeers().size();
+        LOG.info("numFromHADetails {}, numFromRatisServer {}",
+            numFromHADetails, numFromRatisServer);
+        if (handler.rotationPrepareAcks() == numFromRatisServer) {
+          // all acks are received.
+          try {
+            waitAckTimeoutTask.cancel(false);
+            handler.rotationCommit(rootCACertId);
+            handler.rotationCommitted(rootCACertId);
+
+            metrics.incrSuccessRotationNum();
+            long timeTaken = System.nanoTime() - processStartTime.get();
+            metrics.setSuccessTimeInNs(timeTaken);
+            processStartTime.set(null);
+
+            // reset state
+            handler.resetRotationPrepareAcks();
+            String msg = "Root certificate " + rootCACertId +
+                " rotation is finished successfully after " + timeTaken + " ns";
+            cleanupAndStop(msg);
+          } catch (Throwable e) {
+            LOG.error("Execution error", e);
+            handler.resetRotationPrepareAcks();
+            cleanupAndStop("Execution error, " + e.getMessage());
+          } finally {
+            // stop this task to re-execute again in next cycle
+            throw new RuntimeException("Exit the this " +
+                "WaitSubCARotationPrepareAckTask for root certificate " +
+                rootCACertId + " since the rotation is finished execution");
+          }
+        }
+      }
+    }
+  }
+
   /**
    * Stops scheduled monitor task.
    */
   @Override
   public void stop() {
+    if (metrics != null) {
+      metrics.unRegister();
+    }
     try {
       executorService.shutdown();
       if (!executorService.awaitTermination(3, TimeUnit.SECONDS)) {
```

Comment:
As we discussed, due to the various calls to scm.shutDown in case of unrecoverable failures during the rootCA rotation process, this wait here is not fortunate, as there is a circular dependency here.

Once we call scm.shutDown, it will call into the stop method of the manager, then wait 3 seconds to see if the manager stops, but that fundamentally blocks the manager, and shutdown is graceful, so there won't be interruptions either.

I think we should call the shutdownNow() method only on the executorService, and whenever we call scm.shutDown also bail out from further execution either by throwing an exception, or by returning to the caller after scm.shutDown is called. This way we can ensure that scm shuts down properly, and none of the rest of the execution that might happen in the executor puts any unexpected garbage into the on-disk structure, while the executor also tears down gracefully.

The rollbacks and cleanups seem to happen properly, but I am worried what happens if we leave the executor to run the thing further, especially because the executorService uses daemon threads. What do you think?

Reply:
I written one test case to test the effect of  System.exit(), it will not continue execute the statement after it.  So don't worry that the daemon will continue the execution. 

public class TestExecutorService {
  static final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1,
      new ThreadFactoryBuilder().setNameFormat("Test")
          .setDaemon(true).build());

  static void test() throws InterruptedException {
    executor.submit(new Runnable() {
      @Override
      public void run() {
        while(true) {
          if (Thread.currentThread().isInterrupted()) {
            System.out.println("interrupted " + new Date());
            //break;
          } else {
            System.out.println("not interrupted " + new Date());
          }
          executor.shutdownNow();
          System.out.println("shutdownNow is called ");
          System.exit(1);
          System.out.println("exit is called ");
        }
      }
    });
    Thread.sleep(1000);
    System.out.println("Exiting normally...");
  }

  public static void main(String[] args) {
    try {
      test();
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

And shutdownNow is a good advice.

Reply:
Right, let's use scm.shutdown in combination with just shutdownNow in the scm.stop() method for the rotation manager. Thank you for looking into it!

Reply:
Right, let's use scm.shutdown in combination with just shutdownNow in the scm.stop() method for the rotation manager. Thank you for looking into it!